python/ray/data/impl/stats.py

```diff
@@ -0,0 +1,239 @@
+                 parent: Optional["DatasetStats"],
+                 stats_actor=None):
+        """Create dataset stats.
+
+        Args:
+            stages: Dict of stages used to create this Dataset from the
+                previous one. Typically one entry, e.g., {"map": [...]}.
+            parent: Reference to parent Dataset's stats.
+            stats_actor: Reference to actor where stats should be pulled
+                from. This is only used for Datasets using LazyBlockList.
+        """
+
+        self.stages: Dict[str, List[BlockMetadata]] = stages
+        self.parent: Optional["DatasetStats"] = parent
+        self.number: int = 0 if not parent else parent.number + 1
+        self.dataset_uuid: str = None
+        self.time_total_s: float = 0
+        self.stats_actor = stats_actor
+
+        # Iteration stats, filled out if the user iterates over the dataset.
+        self.iter_wait_s: float = 0
+        self.iter_process_s: float = 0
+        self.iter_user_s: float = 0
+        self.iter_total_s: float = 0
+
+    def child_builder(self, name: str) -> _DatasetStatsBuilder:
+        """Start recording stats for an op of the given name (e.g., map)."""
+        return _DatasetStatsBuilder(name, self)
+
+    def child_TODO(self, name: str) -> "DatasetStats":
+        """Placeholder for child ops not yet instrumented."""
+        return DatasetStats(stages={name + "_TODO": []}, parent=self)
+
+    @staticmethod
+    def TODO():
+        """Placeholder for ops not yet instrumented."""
+        return DatasetStats(stages={"TODO": []}, parent=None)
+
+    def summary_string(self, already_printed: Set[str] = None) -> str:
+        """Return a human-readable summary of this Dataset's stats."""
+
+        if self.stats_actor:
+            # XXX this is a super hack, clean it up
+            stats_map, self.time_total_s = ray.get(
+                self.stats_actor.get.remote())
+            for i, metadata in stats_map.items():
+                self.stages["read"][i] = metadata
+        out = ""
+        if self.parent:
+            out += self.parent.summary_string(already_printed)
+            out += "\n"
+        for stage_name, metadata in self.stages.items():
+            out += "Stage {} {}: ".format(self.number, stage_name)
+            if already_printed and self.dataset_uuid in already_printed:
+                out += "[execution cached]"
+            else:
+                if already_printed is not None:
+                    already_printed.add(self.dataset_uuid)
+                out += self._summarize_blocks(metadata)
+        out += self._summarize_iter()
+        return out
+
+    def _summarize_iter(self) -> str:
+        out = ""
+        if self.iter_total_s or self.iter_wait_s or self.iter_process_s:
+            out += "\nDataset iterator time breakdown:\n"
+            out += "* In ray.wait(): {}\n".format(fmt(self.iter_wait_s))
+            out += "* In format_batch(): {}\n".format(fmt(self.iter_process_s))
```

Comment:
Feel its a bit confusing, if it's just format_batch, maybe rename to iter_format_batch_s?

Reply:
Done.